File: learn/graph/prim/prim.go


import (
	"container/heap"
	"math"
)

type PrimEdge struct {
	to     int
	weight int
}

type PrimItem struct {
	vertex int
	weight int
	index  int
}

type PrimPQ []*PrimItem

func (pq PrimPQ) Len() int           { return len(pq) }
func (pq PrimPQ) Less(i, j int) bool { return pq[i].weight < pq[j].weight }
func (pq PrimPQ) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}
func (pq *PrimPQ) Push(x interface{}) {
	n := len(*pq)
	item := x.(*PrimItem)
	item.index = n
	*pq = append(*pq, item)
}
func (pq *PrimPQ) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

type Prim struct {
	vertices int
	adjList  [][]PrimEdge
}

func NewPrim(v int) *Prim {
	return &Prim{
		vertices: v,
		adjList:  make([][]PrimEdge, v),
	}
}

func (p *Prim) AddEdge(u, v, weight int) {
	p.adjList[u] = append(p.adjList[u], PrimEdge{to: v, weight: weight})
	p.adjList[v] = append(p.adjList[v], PrimEdge{to: u, weight: weight})
}

func (p *Prim) MST() int {
	key := make([]int, p.vertices)
	inMST := make([]bool, p.vertices)

	for i := range key {
		key[i] = math.MaxInt32
	}
	key[0] = 0

	pq := make(PrimPQ, 0)
	heap.Init(&pq)
	heap.Push(&pq, &PrimItem{vertex: 0, weight: 0})

	mstWeight := 0

	for pq.Len() > 0 {
		item := heap.Pop(&pq).(*PrimItem)
		u := item.vertex

		if inMST[u] {
			continue
		}

		inMST[u] = true
		mstWeight += item.weight

		for _, edge := range p.adjList[u] {
			v := edge.to
			weight := edge.weight

			if !inMST[v] && key[v] > weight {
				key[v] = weight
				heap.Push(&pq, &PrimItem{vertex: v, weight: key[v]})
			}
		}
	}

	return mstWeight
}

File: learn/graph/prim/prim.rs
use std::cmp::Ordering;
use std::collections::BinaryHeap;

#[derive(Copy, Clone, Eq, PartialEq)]
pub struct PrimState {
    pub cost: i32,
    pub position: usize,
}

impl Ord for PrimState {
    fn cmp(&self, other: &Self) -> Ordering {
        other
            .cost
            .cmp(&self.cost)
            .then_with(|| self.position.cmp(&other.position))
    }
}

impl PartialOrd for PrimState {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

pub struct PrimEdge {
    pub to: usize,
    pub weight: i32,
}

pub struct Prim {
    pub vertices: usize,
    pub adj_list: Vec<Vec<PrimEdge>>,
}

impl Prim {
    pub fn new(vertices: usize) -> Self {
        Prim {
            vertices,
            adj_list: (0..vertices).map(|_| Vec::new()).collect(),
        }
    }

    pub fn add_edge(&mut self, u: usize, v: usize, weight: i32) {
        self.adj_list[u].push(PrimEdge { to: v, weight });
        self.adj_list[v].push(PrimEdge { to: u, weight });
    }

    pub fn mst(&self) -> i32 {
        let mut key: Vec<_> = (0..self.vertices).map(|_| i32::MAX).collect();
        let mut in_mst = vec![false; self.vertices];

        let mut heap = BinaryHeap::new();

        key[0] = 0;
        heap.push(PrimState {
            cost: 0,
            position: 0,
        });

        let mut mst_weight = 0;

        while let Some(PrimState { cost, position }) = heap.pop() {
            if in_mst[position] {
                continue;
            }

            in_mst[position] = true;
            mst_weight += cost;

            for edge in &self.adj_list[position] {
                if !in_mst[edge.to] && edge.weight < key[edge.to] {
                    key[edge.to] = edge.weight;
                    heap.push(PrimState {
                        cost: edge.weight,
                        position: edge.to,
                    });
                }
            }
        }

        mst_weight
    }
}
